### PR TITLE
Fix #1.

### LINKED ISSUE #1: Security Oversights
I should preface this issue by saying that I'm not a member of the Minecraft seed cracking community and don't have an in-depth understanding of Minecraft worldgen, although I do have extensive experience seed cracking other games.

According to my brief browsing of the code, this mod computes a comparatively expensive hash of the world seed once on world generation, then uses the hash as a virtual world seed when needed in some specific functions relevant to cracking. There's a big issue with this though - the virtual seed and the real seed are in practice the same information for those specific functions after startup, as the virtual seed never changes, so knowing the virtual seed doesn't tell you about the parts of the seed using the original seed but critically still _does_ tell you about all of the other generation using the virtual seed, which you now know.

This is a big problem, given that chest loot, structure locations, end pillar layout, etc. are probably useful information to have to a nefarious runner, even if other things like biomes, trees, and cacti of varying size are different. The original tool would be able to obtain this information even without any modification. We would want to find a different way to obfuscate these values. The most obvious and direct solution would be to, for example, use a set of virtual seeds sampled after 32, 64, 96, etc. successive hash applications, one for each mixin, or some similar scheme so that the results of each type of generation can't be compared with each other, rendering the technique useless.

This still isn't completely secure if the specific relationship between hashes is known and a novel cracker that could uniquely identify a seed all on its own (excluding information from the protected generation types) is produced, in which case we would still be able to get all of the hashes anyway. However, you made it sound in the video like that was not a likely scenario as other generation types were too expensive to reverse in real time. If this turns out to still be a problem, it would be trivial to make the relationship between the original world seed and the virtual seed passed to each mixin also randomized, which should be suitably secure. To be honest, we might also want to do away with the xoshiro iterations entirely and use a cryptographically secure hash to more completely thwart the reversibility of the function.

### PR DESCRIPTION
**Pull request by pudy248**
Implements changes suggested in issue #1. I haven't bothered to check if this compiles or runs or anything so probably do that before merging. Compiling java is a pain for a poor C++ developer like myself.

### CODE REVIEW

File: src/main/java/com/kludwisz/anticracker/SeedHash.java [hunk 1]
```diff
@@ -31,13 +31,16 @@ public static void precalculateWorldSeedHash(long worldSeed) {
             return;
 
         long hsh = worldSeed;
+        hashes = new int[NUM_HASHES];
```
Comment (Kludwisz) [inline src/main/java/com/kludwisz/anticracker/SeedHash.java]: static final fields can't be initialized here, this should be moved to the field declaration, also need to change int to long

File: src/main/java/com/kludwisz/anticracker/SeedHash.java [hunk 2]
```diff
@@ -65,9 +68,11 @@ private static long xrsr128pp(long seed) {
...
 
     /**
+     * @param index The index of the hash to use.
      * @return The precalculated hash of the world seed.
      */
-    public static long getWorldSeedHash() {
-        return result.get();
+    public static long getWorldSeedHash(int index) {
+        // if (index > NUM_HASHES) halt_and_catch_fire();
+        return result[index];
```
Comment (Kludwisz) [inline src/main/java/com/kludwisz/anticracker/SeedHash.java]: need to change name to hashes

File: src/main/java/com/kludwisz/anticracker/mixin/ChunkRandomMixin.java [hunk 3]
```diff
@@ -13,21 +13,21 @@
...
-        long scrambledSeed = cir.getReturnValue() ^ SeedHash.getWorldSeedHash();
+        long scrambledSeed = cir.getReturnValue() ^ SeedHash.getWorldSeedHash(3);
         setSeed(scrambledSeed);
         cir.setReturnValue(scrambledSeed);
     }
 
     @Inject(at = @At("RETURN"), method = "setCarverSeed", cancellable = true)
     private void carverseedRet(CallbackInfoReturnable<Long> cir) {
-        long scrambledSeed = cir.getReturnValue() ^ SeedHash.getWorldSeedHash();
+        long scrambledSeed = cir.getReturnValue() ^ SeedHash.getWorldSeedHash(2);
```
Comment (Kludwisz) [inline src/main/java/com/kludwisz/anticracker/mixin/ChunkRandomMixin.java]: would be better to use a different index than the one for population seeding, need to increase the number of hashes to at least 5

**Review by Kludwisz — CHANGES_REQUESTED** [review]
These changes definitely will improve security, there are a couple of technical issues that need to be addressed first, though.

**Comment by pudy248** [comment]
Some would say you shouldn't write code in the Github text editor at 1am in a language you only vaguely recall the specific semantics of.

**Pull request merged**